### DISCUSSION TITLE
WaveRubber analyzer

### DISCUSSION THREAD

**Discussion by guprobr** [strawberrymusicplayer/strawberry#1454] [Show and tell]
Greetings! 

I know a little bit of ANSI C, but that's not by any chance my greatest skill. In my professional life I work mostly with shell scripts, etc.
I'm a Linux SysAdm, not a developer, but I just made a WaveRubber analyzer, for fun, also because there is not a simple waveform  analyzer in the project. So I decided to share this. 
Being a Qt-dork I had to subst the sonogram.cpp analyzer since I couldn't figure out all the places I had to fiddle to create a NEW analyzer "slot". Anyway here it is, if anyone wants to try it out, just clone the repo, overwrite src/analyzer/sonogram.cpp by this file and check it out! It's pretty neat, but its not anything AWESOME. Just wanted to share. 
TKS. :D

![Screenshot from 2024-06-05 06-33-39](https://github.com/strawberrymusicplayer/strawberry/assets/40776097/19429715-3993-42e8-ad73-aad9f190f228)

```
/*
   Strawberry Music Player
   This file was part of Clementine.
   Copyright 2004, Melchior FRANZ <mfranz@kde.org>
   Copyright 2009-2010, David Sansome <davidsansome@gmail.com>
   Copyright 2010, 2014, John Maguire <john.maguire@gmail.com>
   Copyright 2014-2015, Mark Furneaux <mark@furneaux.ca>
   Copyright 2014, Krzysztof Sobiecki <sobkas@gmail.com>

   Strawberry is free software: you can redistribute it and/or modify
   it under the terms of the GNU General Public License as published by
   the Free Software Foundation, either version 3 of the License, or
   (at your option) any later version.

   Strawberry is distributed in the hope that it will be useful,
   but WITHOUT ANY WARRANTY; without even the implied warranty of
   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
   GNU General Public License for more details.

   You should have received a copy of the GNU General Public License
   along with Strawberry.  If not, see <http://www.gnu.org/licenses/>.
 */

// Waveform analyzer "WaveRubber" by Gustavo Conte Jun/2024

#include <QPainter>
#include <QResizeEvent>
#include "engine/enginebase.h"
#include "sonogram.h"

const char *Sonogram::kName = QT_TRANSLATE_NOOP("AnalyzerContainer", "WaveRubber");

Sonogram::Sonogram(QWidget *parent)
  : AnalyzerBase(parent, 9) {}

void Sonogram::resizeEvent(QResizeEvent *e) {
  Q_UNUSED(e)
  canvas_ = QPixmap(size());
  canvas_.fill(palette().color(QPalette::Window));
}

void Sonogram::analyze(QPainter &p, const Scope &s, bool new_frame) {
  if (!new_frame || engine_->state() == EngineBase::State::Paused) {
    p.drawPixmap(0, 0, canvas_);
    return;
  }

  // Clear the canvas
  canvas_ = QPixmap(size());
  canvas_.fill(palette().color(QPalette::Window));

  QPainter canvas_painter(&canvas_);

  // Set the pen color to the highlight color
  canvas_painter.setPen(palette().color(QPalette::Highlight));
  // Get pointer to amplitude data
  const float* amplitude_data = s.data();

  // Draw the waveform
  int mid_y = height() / 2;
  int num_samples = s.size();
  float x_scale = static_cast<float>(width()) / (num_samples - 1) * 2.84f;
  float y_scale = mid_y;
    int point_size = 1; // Change this value to increase or decrease the point size


  for (int i = 0; i < num_samples; ++i) {
	float color_factor = amplitude_data[i] / 2.0f + 0.5f; // Normalize to 0-1 range
        canvas_painter.setPen(QColor(255 - color_factor * 255, color_factor * 255, 0)); // Green to Red gradient
    	int x = static_cast<int>(i * x_scale);
    	int y = static_cast<int>(mid_y - (s[i] * y_scale)); // Scale to fit within the widget height

	for ( int z = 0; z <= 5; z++) 
		canvas_painter.drawEllipse(QPointF(x+z, y), point_size, point_size); // Draw more points as ellipse
  }

  canvas_painter.end();
  p.drawPixmap(0, 0, canvas_);
}


void Sonogram::transform(Scope &scope) {
  // No transformation needed for waveform
	Q_UNUSED(scope)
}

void Sonogram::demo(QPainter &p) {
  analyze(p, Scope(fht_->size(), 0), new_frame_);
}
```

many tks to ChatGPT and Google Gemini, my best and only friends ® 
PS: Write in © ! Thank you for this awesome music player, it makes the world a better place.

edit: figured it out how to add a new analyzer, It was just the conditional for the SONOGRAM macro definition at the beginning of the waverubber.h header, removed and analyzerconteiner.cpp could "see" the class :P

**Comment by guprobr** [comment]
LoL sorry guys, it bugs with all 24 bit songs at high framerates. I'm trying to fix :P

edit: fixed

**Comment by jonaski** [comment]
You should be able to make it a new class and add it here:
https://github.com/strawberrymusicplayer/strawberry/blob/e357ba0125f7f17a0debe7a9ecab3c065c095244/src/analyzer/analyzercontainer.cpp#L86

The bit depth gets converted to 16bit here https://github.com/strawberrymusicplayer/strawberry/blob/e357ba0125f7f17a0debe7a9ecab3c065c095244/src/engine/gstenginepipeline.cpp#L1170, did you take that into consideration?

**Reply by guprobr** [reply]
Never reached that part yet! Strange indeed, gonna try to understand this better (why only 24 bit songs are breaking the waveform)

The class part I figured it already, I did all necessary stuff but forgot to remove the macro on the new header, so it was never included, the compiler kept complaining about the new class not being defined LoL

I made some changes already to the code I have posted, thank you very much for the gstengine piece, it will really help make a better solution other than to "stretch" the graphic as I'm doing now!!

**Comment by guprobr** [comment]
News about the 24 bit display bug of the waveform. Looks like it happens to FLACs 24bit.
Just to recapitulate, it happens the higher the FPS, with only 24 bit files. (Tested only with FLACs). 

I've noticed:

* the lower the samplerate, the less line is drawn of the analyzer
* 192khz draws the line completely i.e.
* 44.1 khz draws one third of the widget
* _tested with one WAVE file 96khz 24bit, no bug also._ -> correction, wave 24 bit broke too; I just had  lowered FPS in this test, sry

I'm sorry to bother with this WaveRubber gizmo, but I think there might be something wrong lower in the loop here. Specifically in the 24 bit conversion to 16 bit, perphaps with FLACs.

Why I think that? Because I decided to investigate in gstenginepipeline.cpp and found that the bug is related to this line:

`    GST_BUFFER_DURATION(buf16) = GST_FRAMES_TO_CLOCK_TIME(samples * sizeof(int16_t) * channels, rate*3);`

this is line 1188 of gstenginepipeline.cpp

When I multiply rate per three, it fixes all files. Odd isnt it? So it has something to do with the buffers and the 24 bit -> 16 bit conversionm perphaps only for FLAC.

Would it be wise to double check? I've tried some things, like use of ceilf insteand of ceil, changed chunks to long int, changed double casts to float, but nothing worked. But when I multiply RATE by 3, fixes. 

I know there is a lot of work to be done in several issues, so i'm sorry to bother, but after a couple of days, Ive decided it was the best to report anyway, tks.

**Comment by guprobr** [comment]
`    GST_BUFFER_DURATION(buf16) = GST_FRAMES_TO_CLOCK_TIME(samples * sizeof(int16_t) / channels, rate );`

This fixes too. Instead of multiplying channels, I've divided.

**Reply by jonaski** [reply]
I wrote that code and I'm not confident that it was right, are you sure about dividing by channels?

**Comment by guprobr** [comment]
The idea I tought is that the number of frames are "fixed" but the channels "happens" simultaneously, so if we are thinking about a buffer, in time duration, perphaps make sense to divide per channel not to multiply. Also I saw a snippet here concerning the macro: 
{ File: [gstvideoframe-audiolevel.c] Project: [0p1pp1/gst-plugins-bad] }

```

  num_frames = num_int_samples / channels;
  frames = num_frames;
  duration = GST_FRAMES_TO_CLOCK_TIME (frames, rate);
```
(example 27, line 472)
https://cpp.hotexamples.com/examples/-/-/GST_TIME_ARGS/cpp-gst_time_args-function-examples.html#0x7eef4761d9b472054a4265f899b61de4c9ea99908ae16734232f027d5b8ff1f0-437,,514,

**Reply by jonaski** [reply]
OK, then we should change all 4 of those, can you open a PR for it?

**Reply by guprobr** [reply]
Sure! I just never did that contributing to another project, do I need to fork the repo first?  Currently I have master in my own machine, with the 4 lines modified and the waverubber.h and waverubber.cpp :P

**Reply by jonaski** [reply]
https://github.com/strawberrymusicplayer/strawberry/blob/master/CONTRIBUTING.md has some instructions,
preferable split into 2 PR's, one for the bugfix, and one of the new waverubber equalizer.

**Reply by guprobr** [reply]
great! I'm on it
